### ISSUE TITLE
Error is being re-raised after being handled, sort of

### ISSUE THREAD

**Issue by stakach** [rubinius/rubinius#3706]
## Problem

Fiber.yield -> error raised -> catch error -> \
Fiber.resume -> re-raise error -> handle error -> Fiber ends
Fiber that yielded has the error raised on it

Something like that anyway. I have it captured in a test case.
## How to Reproduce

Probably easiest to do on ubuntu linux
1. Clone: https://github.com/cotag/libcouchbase/ ([commit](https://github.com/cotag/libcouchbase/tree/28993c306c72c97246bf63fed920a16d2ebedfe2))
2. Run the pre-script steps in `.travis.yml`
3. `rspec ./spec/bucket_spec.rb:62` and `rspec ./spec/bucket_spec.rb:101` (the two tests that catch errors)

If I modify bucket_spec.rb:62 to be:

``` ruby
        it "should retry when performing a CAS operation" do
            @reactor.run { |reactor|
                begin
                    @bucket.set('somekey', 'woop woop')
                    result = @bucket.cas('somekey', retry: 2) do |current|
                        @log << current
                        # This ensures the operation fails
                        @bucket.set('somekey', 'woop woop1')
                        "current #{current}"
                    end
                    @log << result.value
                rescue Libcouchbase::Error::KeyExists
                    # Error handled here
                    @log << :error
                end
                @reactor.stop # Added, this shouldn't be required either
            }
            # Error re-raised here - if you catch it, like above, and add @log << :error
            # then @log == ['woop woop', :error, :error]
            expect(@log).to eq(['woop woop', 'woop woop1', 'woop woop1', :error])
        end
```

This is where the error is originally handled: https://github.com/cotag/libcouchbase/blob/28993c306c72c97246bf63fed920a16d2ebedfe2/lib/libcouchbase/bucket.rb#L246

That code sort of functions properly too, however it also seems to suffer from error replay stemming from: https://github.com/cotag/libuv/blob/f4d5d2467452f3c9d157a5d25ab3214e26a15d86/lib/libuv/coroutines.rb#L45 which is re-raised on the new fiber at L59

Basically the error seems to be handled fine however when re-raised on a new fiber it seems to raise itself on the old fiber?
It's reasonably confusing. I'll try and write a more compact piece of code so it's easier to replicate and follow.
## Configuration Details

I'm running rubinius installed on OSX - version rbx-3.62
Also occurs on travis: https://travis-ci.org/cotag/libcouchbase/jobs/170357082

**Comment by stakach** [comment]
A bit more information:
Putting print statements in the connection class confirms that the get succeeds - so an error shouldn't be re-raised where it is and the backtrace confirms that it occurred in set.

```ruby
        def compare_and_swap(key, **opts)
            retries = opts.delete(:retry) || 0
            begin
                current = result(@connection.get(key)) # <-- Error re-raised here on first retry
                new_value = yield current.value, opts
                opts[:cas] = current.cas

                set(key, new_value, **opts)  # <-- Error first raised here - set calls result, used above
            rescue Libcouchbase::Error::KeyExists
                retries -= 1
                retry if retries >= 0
                raise
            end
        end
```

The same error is re-raised a third time on a totally different fiber when it resumes as these fibers complete

**Comment by brixen** [comment]
@stakach thanks for reporting this and all the work on the repro. There is state that tracks when an exception is raised on a Thread, and that needs to be managed on the main Thread that the Fibers are associated with. That state probably isn't getting reset when the Fiber completes.

**Comment by stakach** [comment]
That makes sense. I wrapped [this line](https://github.com/cotag/libuv/blob/f4d5d2467452f3c9d157a5d25ab3214e26a15d86/lib/libuv/coroutines.rb#L53) in a begin / rescue block to confirm that the error is re-raised here.

![image](https://cloud.githubusercontent.com/assets/368013/19710353/7a451e2a-9b79-11e6-8887-8c187921f028.png)

That's the `error in Fiber resume` ouput in that screenshot.
I think it is impossible for that line to raise an error in MRI (unless you are on the root Fiber)

**Comment by stakach** [comment]
Once this is resolved it should also fix #2966 too. I have the code passing all tests on MRI and jRuby now.
https://travis-ci.org/cotag/libcouchbase

**Comment by stakach** [comment]
Found another location where this issue rears it's head, occurs in a less obscure way - might make it easier to build some example code that tests the issues.

https://github.com/cotag/libcouchbase/blob/6e4381dcbc2513bbd6d81678d173409042b124e3/lib/libcouchbase/bucket.rb#L791

Basically the sleep function there starts a timer and yields the fiber - however when the fiber resumes the error is re-raised and hence the block is not retried

**Comment by brixen** [comment]
@stakach thank you! I'll take a look today.

**Comment by stakach** [comment]
Minimal code to reproduce the error

```ruby
require 'libuv'
reactor.run { |reactor|
    tries = 0
    begin
        raise 'bad' if tries == 0
        puts 'recovered'
    rescue => e
        reactor.sleep 200
        tries += 1
        retry if tries == 1
    end
    puts "made it here"
}
```

**Comment by brixen** [comment]
Here's a repro:

```ruby
f = Fiber.new do
  puts "0. fiber starting"

  begin
    puts "1. fiber raising"
    raise "an exception in a Fiber"
  rescue => e
    puts "2. fiber rescuing"
    puts "rescued: #{e.class}, #{e.message}"

    puts "3. fiber yielding"
    Fiber.yield :yielding
  end

  puts "4. fiber exiting"
  :exiting
end

puts ">0. resume value: #{f.resume.inspect}"
puts ">1. resume value: #{f.resume.inspect}"
```

When `Fiber.yield` executes, switching the current Fiber on the Thread, the exception is raised in the Thread.

I cannot conceive of any way that `Fiber.yield` could execute with an unhandled exception, so propagating the exception seems to be the problem.

Exceptions raised and unhandled on a Fiber should be propagated:

```ruby
fiber_running = false

t = Thread.new do
  f = Fiber.new do
    puts "0. fiber starting"

    fiber_running = true

    puts "1. fiber raising"
    raise "an exception in a Fiber"

    puts "2. fiber exiting"
    :exiting
  end

  begin
    puts ">0. resume value: #{f.resume.inspect}"
    puts ">1. resume value: #{f.resume.inspect}"
  rescue => e
    puts ">2. rescuing: #{e.class}, #{e.message}"
  end
end

puts "t0. main thread running"

Thread.pass until fiber_running

puts "t1. joining thread"
puts "t2. value of thread: #{t.value.inspect}"
```

**Comment by stakach** [comment]
Love you @brixen !
Awesome work and many thanks!

**Comment by brixen** [comment]
@stakach thanks! Note that there may be a race in Fiber::cancel_all that I'm looking at right now, but it should be fixed soon. I removed the 3.76 tag for now.

**Comment by brixen** [comment]
@stakach I fixed the issue with Fiber that https://github.com/rubinius/rubinius/commit/c9b05989ec16492b3e34dfca4d1b001c46b85d24 introduced and 3.76 is tagged. Please let me know if you have any issues with it.

**Comment by stakach** [comment]
@brixen all my tests are passing now 👍 thanks again!

**Comment by brixen** [comment]
@stakach great to hear and thank you for opening the issue and creating a repro, it was a big help.
